sgv2-graphqlapi/src/main/java/io/stargate/sgv2/graphql/schema/cqlfirst/dml/fetchers/MutationFetcher.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.sgv2.graphql.schema.cqlfirst.dml.fetchers;
+
+import static io.stargate.sgv2.graphql.schema.SchemaConstants.ASYNC_DIRECTIVE;
+import static io.stargate.sgv2.graphql.schema.SchemaConstants.ATOMIC_DIRECTIVE;
+
+import graphql.GraphQLException;
+import graphql.language.OperationDefinition;
+import graphql.schema.DataFetchingEnvironment;
+import io.stargate.proto.QueryOuterClass.Batch;
+import io.stargate.proto.QueryOuterClass.Query;
+import io.stargate.proto.QueryOuterClass.QueryParameters;
+import io.stargate.proto.Schema.CqlTable;
+import io.stargate.sgv2.common.futures.Futures;
+import io.stargate.sgv2.graphql.schema.cqlfirst.dml.NameMapping;
+import io.stargate.sgv2.graphql.web.resources.StargateGraphqlContext;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public abstract class MutationFetcher extends DmlFetcher<CompletionStage<Map<String, Object>>> {
+
+  protected MutationFetcher(String keyspaceName, CqlTable table, NameMapping nameMapping) {
+    super(keyspaceName, table, nameMapping);
+  }
+
+  @Override
+  protected CompletionStage<Map<String, Object>> get(
+      DataFetchingEnvironment environment, StargateGraphqlContext context) throws Exception {
+
+    Query query = null;
+    Exception buildException = null;
+
+    // Avoid mixing sync and async exceptions
+    try {
+      // buildQuery() could throw an unchecked exception.
+      // As the statement might be part of a batch, we need to make sure the
+      // batched operation completes.
+      query = buildQuery(environment, context);
+    } catch (Exception e) {
+      buildException = e;
```

Comment:
`executeAsPartOfBatch` checks the exception before it tries to do anything with the query.

This fetcher is a bit convoluted because in a query like this:
```
mutation @atomic {
  m1: updatebooks(value: { title: "1984", author: "G.O." } ) { applied },
  m2: insertauthors(value: { author: "G.O.", title: "1984" } ) { applied },
  m3: deletebooks(value: { title: "Animal Farm" } ) { applied }
}
```
m1, m2 and m3 will be handled by 3 different fetcher instances.

If everything goes well, each fetcher will accumulate a query in `executeAsPartOfBatch()`, and the last one to get there will execute the batch at line 102.

But if say m1 fails during `buildQuery()`, we still want to go into `executeAsPartOfBatch()` to store the exception in the batch context, so that it can also be used to fail the two other mutations (it's all or nothing).

Reply:
OK, had no idea that something like that is possible as well..